sql/sql_admission_control.h

```diff
@@ -0,0 +1,356 @@
+  MULTI_TENANCY_RET_ACCEPT = 0,
+  MULTI_TENANCY_RET_WAIT,
+  MULTI_TENANCY_RET_REJECT,
+
+  MULTI_TENANCY_RET_FALLBACK,
+  MULTI_TENANCY_NUM_RETURN_TYPES
+};
+
+typedef enum enum_multi_tenancy_resource_type MT_RESOURCE_TYPE;
+typedef enum enum_multi_tenancy_return_type MT_RETURN_TYPE;
+typedef std::unordered_map<std::string, std::string> ATTRS_MAP_T;
+typedef std::vector<std::pair<std::string, std::string>> ATTRS_VEC_T;
+
+struct multi_tenancy_resource_attributes {
+  const ATTRS_MAP_T *connection_attrs_map;
+  const ATTRS_VEC_T *query_attrs_vec;
+  const char *database;
+};
+
+typedef struct multi_tenancy_resource_attributes MT_RESOURCE_ATTRS;
+
+// The contents here must match entries in admission_control_filter_names array
+enum enum_admission_control_filter {
+  ADMISSION_CONTROL_ALTER,
+  ADMISSION_CONTROL_BEGIN,
+  ADMISSION_CONTROL_COMMIT,
+  ADMISSION_CONTROL_CREATE,
+  ADMISSION_CONTROL_DELETE,
+  ADMISSION_CONTROL_DROP,
+  ADMISSION_CONTROL_INSERT,
+  ADMISSION_CONTROL_LOAD,
+  ADMISSION_CONTROL_SELECT,
+  ADMISSION_CONTROL_SET,
+  ADMISSION_CONTROL_REPLACE,
+  ADMISSION_CONTROL_ROLLBACK,
+  ADMISSION_CONTROL_TRUNCATE,
+  ADMISSION_CONTROL_UPDATE,
+  ADMISSION_CONTROL_SHOW,
+  ADMISSION_CONTROL_USE,
+  ADMISSION_CONTROL_END = 64
+};
+
+/*
+ * sql_admission_control.h/cc
+ *
+ * This module handles multi-tenancy resource allocation on the server side.
+ * Functions will call into multi-tenancy plugin interfaces (if installed) to
+ * make a decision of whether or not a resource can be allocated. Connections
+ * and query limits are among the resources multi-tenancy plugin allocates.
+ * Note that admission control is now part of the multi-tenancy module to
+ * hanlde query throttling.
+ *
+ * The isolation level is defined by an entity. The entity could be a database,
+ * a user, or any thing that multi-tenancy plugin defines to isolate the
+ * resource allocation.
+ *
+ * See sql_admission_control.cc for implementation.
+ */
+
+extern int multi_tenancy_admit_query(THD *, const MT_RESOURCE_ATTRS *);
+extern int multi_tenancy_exit_query(THD *, const MT_RESOURCE_ATTRS *);
+
+/**
+  Per-thread information used in admission control.
+*/
+
+#ifdef HAVE_PSI_INTERFACE
+
+static PSI_mutex_key key_ac_node_lock;
+static PSI_cond_key key_ac_node_cond;
+
+static PSI_mutex_info all_ac_node_mutexes[] = {
+    {&key_ac_node_lock, "st_ac_node::lock", PSI_FLAG_SINGLETON, 0,
+     PSI_DOCUMENT_ME}};
+
+static PSI_cond_info all_ac_node_conds[] = {
+    {&key_ac_node_cond, "st_ac_node::cond", PSI_FLAG_SINGLETON, 0,
+     PSI_DOCUMENT_ME}};
+
+static void init_st_ac_node_keys() {
+  const char *const category = "sql";
+  int count;
+
+  count = static_cast<int>(array_elements(all_ac_node_mutexes));
+  mysql_mutex_register(category, all_ac_node_mutexes, count);
+
+  count = static_cast<int>(array_elements(all_ac_node_conds));
+  mysql_cond_register(category, all_ac_node_conds, count);
+}
+#endif
+
+struct st_ac_node {
```

Comment:
make `st_ac_node` non-copyable
```suggestion
struct st_ac_node {
  st_ac_node(const st_ac_node&) = delete;
  st_ac_node& operator = (const st_ac_node&) = delete;
```